### ISSUE TITLE
Please add ALL4 Channels and Catch-Up TV (UK Channel 4 Network, formerly known as 4OD)

### ISSUE THREAD

**Issue by kevwag** [Catch-up-TV-and-More/plugin.video.catchuptvandmore#767]
### Content description
ALL4 is the streaming platform for the UK's Channel 4 Network.
It includes [Catch-up TV](https://www.channel4.com/categories), [Box Sets](https://www.channel4.com/categories/boxsets), and five live channels with [EPG](https://www.channel4.com/tv-guide):

- [Channel 4](https://www.channel4.com/now/C4?intcmp=header_watch_live_c4)
- [E4](https://www.channel4.com/now/E4?intcmp=header_watch_live_e4)
- [More 4](https://www.channel4.com/now/M4?intcmp=header_watch_live_m4)
- [4/7](https://www.channel4.com/now/4S?intcmp=header_watch_live_4s)
- [Film 4](https://www.channel4.com/now/F4?intcmp=header_watch_live_f4)

* Name: ALL4
* Type [e.g. TV channel, Catch-up TV, Website, Radio]: TV Channels (x5) / Catch-up TV
* Website URL: https:www.channel4.com
* Content language/country: English / United Kingdom

### Content access

- [x] This content or some part of this content require a free account
- [ ] This content or some part of this content require a paid account
- [x] This content is geographically restricted

### More information

N/A

**Comment by joaopa00** [comment]
DO you know the api of this site?

**Comment by kevwag** [comment]
> Do you know the api of this site?

[](http://developer.channel4.com/io-docs)

**Comment by joaopa00** [comment]
Here is a beginning of an attempt to implement Channel4 replay: https://github.com/joaopa00/Tests/blob/main/channel4.py

But I a, stuck to retrieve request_id and message in payload. If someone has an idea, please share it !!

**Comment by kevwag** [comment]
@joaopa00  
Thanks for doing this, it'll be awesome to have a bit more UK functionality :)
(I think there's an error in your link URL - I get a 404 page)

**Comment by joaopa00** [comment]
Sorry, the file is in my private github. Here is :
[test.txt](https://github.com/Catch-up-TV-and-More/plugin.video.catchuptvandmore/files/8682482/test.txt)

**Comment by kevwag** [comment]
@joaopa00 
Am I able to paste that file into the addon on my system and try it out or does it need more work first?

**Comment by joaopa00** [comment]
For now, this file fetches the title of the episodes only. It can not play any stream.

Recentle, I had some progress in playing streams. But I still have problems to fetch the token.

**Comment by kevwag** [comment]
> For now, this file fetches the title of the episodes only. It can not play any stream.
> 
> Recently, I had some progress in playing streams. But I still have problems to fetch the token.

Ah, ok.
I remember that the recent fixes for UKTV Play involved hard-coding the token instead of scraping for it. 
Is that maybe an option or is this different?

**Comment by joaopa00** [comment]
It is different. It changes all the time.

**Comment by kevwag** [comment]
Well that sucks then! 
Is there anything in the online API documentation about how to do it?
There seemed to be quite a lot of documentation at the API URL that I posted but I was instantly out of my depths and didn't have a clue what any of it meant!

**Comment by joaopa00** [comment]
On their site, I never managed to fetch a key. So I was never able to  study their API. DO you know how to request a key?.

**Comment by kevwag** [comment]
Once you've signed up for a developer account at [developer.channel4.com](http://developer.channel4.com) and you're logged in, the page to request an API Key is at [developer.channel4.com/apps](http://developer.channel4.com/apps).

I noticed that on the [interactive documentation page](http://developer.channel4.com/io-docs) there are a bunch of different calls and methods for performing various functions and tasks. You can interactively try them all out and although the response is always "forbidden" without a valid API key entered, it does give you the URL construction format and other info that might be useful.

**Comment by joaopa00** [comment]
I registered but received no key...


Channel 4 API Application Form
Application Registered!

To get started using your API keys, dig into our documentation and feel free to ask questions in our forum. We look forward to seeing what you create!

**Comment by kevwag** [comment]
@joaopa00 
Did you check on the [developer.channel4.com/apps/mykeys](http://developer.channel4.com/apps/mykeys) and [developer.channel4.com/apps/myapps](http://developer.channel4.com/apps/mykeys) pages?

The application form suggests that they manually approve key requests rather than automatically issuing a key like many other sites do. 
I found a couple of posts in the developer portal forum from other users who say they didn't receive a key. Those posts are a few years old but never received replies so I don't think anyone's actively monitoring the portal.
I'm guessing that since the API is managed by a third-party company (Tibco), they probably don't want to pay for staff to monitor the portal for API requests or forum posts and most likely don't check the site at all unless the customer (Channel 4) asks them to do something.

I found an email address for Tibco support: [support@tibco.com](mailto:support@tibco.com) on an email confirming that I'd changed my dev account password. It's probably worth dropping them an email saying that you didn't receive a key because they probably won't check the site for a long time unless someone gives them a gentle nudge in the right direction.

**Comment by nictjir** [comment]
Trying to get the channel4 replay to work based on the thread at
https://github.com/xbmc/inputstream.adaptive/issues/995

Where did you find the vale for the basic auth string?
Proxy-Authorization=Basic cDluOGl6YWctcmh3dXl6OTpjamJ5NTZ0dTc5

I have looked through hours of developer tools outputs and output from chrome://net-export/
and can't see where the basic auth string comes from?


any pointers will be a great help to try and help me get this to work.

**Comment by nictjir** [comment]
Thankyou for sharing your code and ideas.
I am also out of the uk :(
Will try and get it to work

**Comment by joaopa00** [comment]
Thank you for your hard work. However, I had already pushed a seemly cleaner solution in Git.

**Comment by nictjir** [comment]
Indeed way way way cleaner! thankyou!

**Comment by joaopa00** [comment]
And now, the lives are implemented in git.
Time to close this issue.

**Comment by Arathonk** [comment]
> And now, the lives are implemented in git. Time to close this issue.

What are the steps? Because I don't really understand what to do, I have the Catch-up Tv & More addon installed but there's no Channel4/All4.

**Comment by joaopa00** [comment]
For now, you need to install the beta version. Have a look at https://github.com/Catch-up-TV-and-More/plugin.video.catchuptvandmore
When the stable version is released, you will be able to install it

**Comment by Arathonk** [comment]
> For now, you need to install the beta version. Have a look at https://github.com/Catch-up-TV-and-More/plugin.video.catchuptvandmore When the stable version is released, you will be able to install it

Yes found it, unfortunately it's not working, when going to Channel 4, and click any show, it just doesn't want to load the show.

**Comment by joaopa00** [comment]
Do you have UK IP?

**Comment by Arathonk** [comment]
> Do you have a UK IP?

Yes I do

**Comment by joaopa00** [comment]
Please attach the log from Kodi here (as a text file)

**Comment by kevwag** [comment]
@joaopa00 
Sorry for making the request and then seemingly disappearing, I've had a lot going on lately and haven't had time to keep up with everything. I did notice the channel 4 section appear in the beta so first of all I'd like to thank you for your hard work, especially with the IPTV integration, it's great that we're getting closer to restoring full Freeview functionality in Kodi without the need for DVB tuners.

I've noticed a bit of a bug in the on-demand section:
It's only recognising and playing streams that are TV show episodes that are in numbered season folders. Movies, shows that only have a single season, One-off's and special episodes don't work, it's as if Kodi is trying to open the url as a folder listing instead of a stream.

Example:
Browse to Catch-up>>United Kingdom>>Channel 4>>Comedy>>Bad Robots: Comedy Blap
This show only has one season, selecting it produces error: _RuntimeError>>no items found_
The same error happens when trying to play any movie or on-off show directly from the comedy menu.

Special episodes of a TV show produce a slightly different error:

Example:
Browse to Catch-up>>United Kingdom>>Channel 4>>Comedy>>The Windsors
There are three season folders titled "Series 1", "Series 2", "Series 3" and also folders titled "Christmas Special" and "Royal Wedding Special". The three season folders work fine but selecting either of the specials produces the error: _IndexError>>list out of range_, I'm guessing that's because it's trying to open the url as a list of folder contents when it should be playing the url as a stream.

Other than this bug, everything else seems good. The epg data's missing for the live channels but that's a known issue affecting all the UK channel listings. The issue has its own error report but that's been open for months without a single response, so I don't know if it'll ever get fixed.
